ner_v2/property_assignor/property_assignor.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import absolute_import
+
+import regex as re  # We are using regex because it provides accurate word boundaries than built in re module
+
+import ner_v2.detectors.utils as v2_utils
+from ner_v2.property_assignor.fields import AssignedProperty
+from ner_v2.property_assignor.reader import PropertiesDataReader
+from ner_v2.property_assignor.utils import clean_text
+
+
+class PropertyAssignor(object):
+    """
+    This module assigns arbitrary properties to already detected entities. Say for example, once you have
+    detected cities, you might want to assign some roles such as "arrival", "departure", "via"
+    etc. based on thier contexts.
+
+    This module takes in
+        - text
+        - spans that are substrings of the text that were detected as some entity
+        - data file that contains context data for each property, such as what should appear before and after entity
+          mention, what should appear in the bot message (it if is given) for this property to be assigned
+
+    Attributes:
+        _property_context_patterns (list): list containing PropertyContextPatterns items.
+    """
+
+    def __init__(self, properties_data_path, subset=None):
+        # type: (str, List[str]) -> None
+        """
+        Args:
+            properties_data_path (str): path to the csv file containing the properties data
+            subset (list of str, optional): list of properties to assign from. Defaults to None, in that case all
+                properties read from the data file are considered to be assigned from.
+        """
+        self._property_context_patterns = PropertiesDataReader.from_csv(  # type: List[PropertyContextPatterns]
+            csv_file_path=properties_data_path,
+            include_only=subset
+        )
+
+    def assign_properties(self, text, detected_spans, bot_message=None):
+        # type: (Text, List[Text], Optional[Text]) -> List[AssignedProperty]
+        """
+        Assign properties to `detected_spans` based on their contexts
+
+        Rough pseudocode:
+            - Replace all detected spans in text with an entity placeholder
+            - Drop punctuations, Reduce whitespace from text and bot_message
+            - If `detected_spans` has only one item, matches are allowed throughout the text instead
+                of being immediately around the detected entity
```

Comment:
Can we make it a bit more clear that in this sentence that in any case where detected span contains multiple items then only immediate presence is considered.